pages/base.py

```diff
@@ -15,89 +17,200 @@ def __init__(self, driver):
+        'logout_link': (By.CSS_SELECTOR, '#secondary-navigation > ul > li.dropdown.open > ul > li:nth-child(4) > a'),
+        'current_service': (By.CSS_SELECTOR, '#navbarScope > div > div > div.service-home > a > span.current-service > strong')
+    }
+
+    def verify(self):
+        return len(self.find_elements(By.ID, 'navbarScope')) == 1
+
+    def is_logged_in(self):
+        try:
+            if self.sign_up_button:
+                return False
+        except ValueError:
+            return True
+
+
+class LoginPage(BasePage):
+    url = settings.OSF_HOME + '/login'
+
+    locators = {
+        'identity': (By.XPATH, '/html/body[@id="cas"]/div[@id="container"]', settings.LONG_TIMEOUT),
+        'username_input': (By.ID, 'username'),
+        'password_input': (By.ID, 'password'),
+        'submit_button': (By.NAME, 'submit'),
+        'remember_me_checkbox': (By.ID, 'rememberMe'),
+    }
+
+    if 'localhost:5000' in settings.OSF_HOME:
+        locators['submit_button'] = (By.ID, 'submit')
+        locators['identity'] = (By.ID, 'login')
+
+    def __init__(self, driver, verify=False):
+        super(LoginPage, self).__init__(driver)
+        if verify:
+            self.check_page()
+
+    def error_handling(self):
+        if self.url not in self.driver.current_url:
+            raise LoginError(
+                driver=self.driver,
+                error_info='Already logged in'
+            )
+
+    def login(self, user, password):
+        self.username_input.send_keys(user)
+        if ('localhost:5000'  not in settings.OSF_HOME):
+            self.password_input.send_keys(password)
+            if self.remember_me_checkbox.is_selected():
+                self.remember_me_checkbox.click()
+        self.submit_button.click()
+
+
+def login(osf_page, user=settings.USER_ONE, password=settings.USER_ONE_PASSWORD):
+    login_page = LoginPage(osf_page.driver)
+    login_page.goto()
+    login_page.login(user, password)
+    osf_page.driver.get(osf_page.url)
+
 
 class OSFBasePage(BasePage):
     url = settings.OSF_HOME
 
-    def __init__(self, driver):
+    # all page must have unique identity
+    locators = {}
+
+    def __init__(self, driver, verify=False, require_login=False):
         super(OSFBasePage, self).__init__(driver)
-        self.navbar = self.Navbar(driver)
 
-    locator_dictionary = {
-        'sign_in_button': (By.LINK_TEXT, 'Sign In')
-    }
+        self.navbar = self.BasePageNavbar(driver)
 
-    def is_logged_in(self):
-        return self.navbar.logged_in()
+        if require_login:
+            self.driver.get(self.url)
+            if not self.is_logged_in():
+                login(self)
 
+        if verify:
+            self.check_page()
 
-    class Navbar(BaseElement):
+    @property
+    def error_heading(self):
+        try:
+            error_head = self.find_element(By.CSS_SELECTOR, 'h2#error')
+        except NoSuchElementException:
+            return None
```

Comment:
Is this enough information when tests are being run if something goes wrong?

Reply:
This is just to catch any error alert happening on OSF. Like "Cannot fork a project. Please contact 'support@osf.io' such kind of error message.

Reply:
Is the expectation that this is running opportunistically on all pages just in case an error shows, or is this really expecting to hit one of those errors? If the latter, and it doesn't hit an error, will there be enough information from the test results to know for sure what happened? I expect so, but just making sure.